src/driftpy/dlob/DLOB.py

```diff
@@ -0,0 +1,1067 @@
+        elif is_variant(record.action, 'FILL'):
+            if record.taker is not None:
+                taker_order = self.get_order(record.taker_order_id, record.taker)
+                if taker_order is not None:
+                    self.update_order(taker_order, record.taker, slot, record.taker_order_cumulative_base_asset_amount_filled)
+
+            if record.maker is not None:
+                maker_order = self.get_order(record.maker_order_id, record.maker)
+                if maker_order is not None:
+                    self.update_order(maker_order, record.maker, slot, record.maker_order_cumulative_base_asset_amount_filled)
+        elif is_variant(record.action, 'CANCEL'):
+            if record.taker is not None:
+                taker_order = self.get_order(record.taker_order_id, record.taker)
+                if taker_order is not None:
+                    self.delete(taker_order, record.taker, slot)
+
+            if record.maker is not None:
+                maker_order = self.get_order(record.maker_order_id, record.maker)
+                if maker_order is not None:
+                    self.delete(maker_order, record.maker, slot)
+    
+    def _get_best_node(
+        self,
+        generator_list: List[Generator[DLOBNode, None, None]],
+        oracle_price_data: OraclePriceData,
+        slot: int,
+        compare_fcn: Callable[[DLOBNode, DLOBNode, int, OraclePriceData], bool],
+        filter_fcn: Optional[DLOBFilterFcn] = None
+    ) -> Generator[DLOBNode, None, None]:
+        # curframe = inspect.currentframe()
+        # calframe = inspect.getouterframes(curframe, 2)
+        # print('caller name:', calframe[1][3])
+        # print(compare_fcn)
+        generators = [{'next': next(generator, None), 'generator': generator} for generator in generator_list]
+
+        side_exhausted = False
+        while not side_exhausted:
+            best_generator = None
+            for current_generator in generators:
+                if current_generator['next'] is None:
+                    continue
+                if best_generator is None or compare_fcn(best_generator['next'], current_generator['next'], slot, oracle_price_data):
+                    best_generator = current_generator
+
+            if best_generator and best_generator['next']:
+                # Skip this node is it's already completely filled or fails filter function
+                if best_generator['next'].is_base_filled() or (filter_fcn and not filter_fcn(best_generator['next'])):
+                    best_generator['next'] = next(best_generator['generator'], None)
+                    continue
+
+                yield best_generator['next']
+                try:
+                    best_generator['next'] = next(best_generator['generator'])
+                except StopIteration:
+                    best_generator['next'] = None
+            else:
+                side_exhausted = True
+
+    def _estimate_fill_exact_base_amount_in_for_side(
```

Comment:
nit: it's nice to put this function under `estimate_fill_with_exact_base_amount`. when a dev is reading `estimate_fill_with_exact_base_amount`, it's more intuitive to scroll down (than up) to find the dependent function